### PR TITLE
feat: add admin form

### PR DESCRIPTION
**Pull request by fedasevich**
## Describe your changes

- Created admin form to create and change admins.
- Added change and create admin routes.
- Added generate password helper.
- Added alerts about copying and regenerating password.

## Issue ticket code (and/or) and link

- [Link to JIRA ticket](https://homecareaid.atlassian.net/browse/CC-236?atlOrigin=eyJpIjoiYzBjNjQ4ODJhNmU4NGJjM2JlODgzMjhmZDYyZjU4ZGMiLCJwIjoiaiJ9)
image
image
image

### **General**

- [x] Assigned myself to the PR
- [ ] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [x] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [x] Date and time formats are on the constants
- [x] Functions are public only if it's used outside the class
- [x] No hardcoded values
- [ ] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Frontend

- [x] Components and business logic are separated
- [x] Colors, Font Size, and Font Name is on the theme or in the constants
- [x] No text in the components, use i18n approach
- [x] No inline styles
- [x] Imports are absolute
- [x] Attach a screenshot if PR has visual changes.

### CODE REVIEW

File: src/components/admin-management/admin-form/AdminForm.tsx [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+                  <Controller
+                    name="phoneNumber"
+                    control={control}
+                    render={({ field, fieldState }): JSX.Element => (
+                      <MuiTelInput
+                        {...field}
+                        onChange={(value: string): void => {
+                          const currentValue = getValues('phoneNumber');
+                          field.onChange(value);
+                          if (value.slice(2, 3) === '0') {
```
Comment (ProggerOleg) [inline src/components/admin-management/admin-form/AdminForm.tsx:162]: magic number

File: src/components/admin-management/admin-form/AdminForm.tsx [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+                            field.onChange('');
+                            setError('phoneNumber', {
+                              type: 'manual',
+                              message: `${translate('adminManagement.adminForm.phoneInvalid')}`,
+                            });
+
+                            return;
+                          }
+                          if (value.slice(2, 3) === '') {
```
Comment (ProggerOleg) [inline src/components/admin-management/admin-form/AdminForm.tsx:171]: same

**Review by ProggerOleg — APPROVED** [review]

File: src/components/admin-management/admin-form/AdminForm.tsx (hunk 1 above)
Comment (fedasevich) [reply src/components/admin-management/admin-form/AdminForm.tsx:162]: fixed

File: src/hooks/useCopyToClipboard.ts [hunk 3]
```diff
@@ -29,12 +32,12 @@ function useCopyToClipboard(): ReturnType {
     } catch (error) {
       // eslint-disable-next-line
       console.log('Copy failed', error);
```
[the remark this answers is not in the document]
Comment (fedasevich) [reply src/hooks/useCopyToClipboard.ts:34]: this is the initial project template code, i think it is better to leave it as it is

**Comment by nadiia-tsytsylina** [comment]
I think "Edit admin" will be better, than "Change" in this context
image

File: src/components/admin-management/admin-form/validation.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+          .transform((originalValue) =>
+            originalValue.trim() === '' ? undefined : originalValue.trim()
+          )
+      : string()
+          .trim()
+          .min(MIN_PASSWORD_LENGTH, translate('adminManagement.adminForm.passwordInvalid')),
+    role: mixed<UserRole>()
+      .oneOf(['Admin'] as UserRole[])
+      .required(),
+    updatedAt: string()
```
Comment (nadiia-tsytsylina) [inline src/components/admin-management/admin-form/validation.ts:56]: Why do you need validation of this field?

Comment (fedasevich) [reply src/components/admin-management/admin-form/validation.ts:56]: it is required to check because role is in form values and needed to create admin

Comment (nadiia-tsytsylina) [reply src/components/admin-management/admin-form/validation.ts:56]: I m about " updatedAt" field

Comment (fedasevich) [reply src/components/admin-management/admin-form/validation.ts:56]: it is to remove it from final request because rhf sends it with other data

**Review by nadiia-tsytsylina — APPROVED** [review]

File: src/components/admin-management/admin-form/AdminForm.tsx [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+              </Grid>
+              <Grid item xs={6}>
+                <FormControl disabled fullWidth variant="filled">
+                  <Controller
+                    name="phoneNumber"
+                    control={control}
+                    render={({ field, fieldState }): JSX.Element => (
+                      <MuiTelInput
+                        {...field}
+                        onChange={(value: string): void => {
```
Comment (Lilly-nn) [inline src/components/admin-management/admin-form/AdminForm.tsx:159]: To be honest I'm not sure if this function should be inline, it's just too large to be directly in the component

File: src/components/admin-management/admin-form/hooks.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  const router = useRouter();
+  const { copiedText, copy, resetCopiedText } = useCopyToClipboard();
+
+  const [showPassword, setShowPassword] = useState<boolean>(false);
+  const [isPasswordRegenerated, setIsPasswordRegenerated] = useState<boolean>(false);
+
+  const [accountCheck] = useAccountCheckMutation();
+
+  const [regeneratePassword] = adminPanelApi.useUpdateAdminPasswordMutation();
+
```
Comment (Lilly-nn) [inline src/components/admin-management/admin-form/hooks.ts:70]: two different code styles, line 67 and 69

File: src/components/admin-management/admin-form/AdminForm.tsx (hunk 5 above)
Comment (fedasevich) [reply src/components/admin-management/admin-form/AdminForm.tsx]: fixed

File: src/components/admin-management/admin-form/hooks.ts (hunk 6 above)
Comment (fedasevich) [reply src/components/admin-management/admin-form/hooks.ts:70]: fixed

**Review by Lilly-nn — APPROVED** [review]

File: src/components/admin-management/admin-form/validation.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    password: undefined;
+  },
+  ''
+> => {
+  const { translate } = useLocales();
+
+  return object({
+    firstName: string()
+      .trim()
+      .max(100, translate('adminManagement.adminForm.firstNameInvalid'))
```
Comment (Simpsonkk) [inline src/components/admin-management/admin-form/validation.ts:28]: 100 - magic number and below the same

Comment (fedasevich) [reply src/components/admin-management/admin-form/validation.ts]: fixed

**Review by Simpsonkk — APPROVED** [review]

**Pull request merged**